util/setuidgid.c

```diff
@@ -0,0 +1,366 @@
+	if (ngroups_max_long < 0) {
+		fprintf(stderr, "Maximum group list size is negative.\n");
+		goto err0;
+	}
+	if (ngroups_max_long >= INT_MAX) {
+		fprintf(stderr, "Maximum group list size is too large.\n");
+		goto err0;
+	}
+	ngroups_max = (int)ngroups_max_long;
+
+	/* Allocate space for group list. */
+	if ((*grouplist = (gid_t *)malloc((unsigned long)ngroups_max *
+	    sizeof(gid_t))) == NULL)
+		goto err0;
+
+	/* Success! */
+	return (ngroups_max);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Check if we're in any supplementary groups. */
+static int
+check_supplementary_groups_none()
+{
+	gid_t *grouplist;
+	int ngroups_max;
+	int ngroups;
+
+	/* Allocate memory for the group list. */
+	if ((ngroups_max = allocate_group_list(&grouplist)) < 0) {
+		warnp("Out of memory");
+		goto err0;
+	}
+
+	/* Get group list. */
+	if ((ngroups = getgroups(ngroups_max, grouplist)) < 0) {
+		warnp("getgroups()");
+		goto err0;
+	}
+
+	/*
+	 * Did we successfully remove ourselves from all groups [1]?
+	 *
+	 * [1] Note that POSIX states that "it is implementation-defined
+	 * whether getgroups() also returns the effective group ID in the
+	 * grouplist array.".
+	 * - We therefore needed grouplist to be large enough (i.e. +1) to
+	 *   hold the _SC_NGROUPS_MAX as well as the effective group ID.
+	 * - We also need to account for the effective group id (egid) if it
+	 *   exists in the array, before judging whether setgroups() was
+	 *   successful or not.
+	 */
+	if (ngroups > 0) {
+		if (!((ngroups == 1) && (grouplist[0] == getegid())))
+			goto err1;
+	}
+
+	/* Clean up. */
+	free(grouplist);
+
+	/* Success! */
+	return (0);
+
+err1:
+	free(grouplist);
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+
+/**
+ * set_group(group_string):
+ * Sets the GID to ${group_string}, which may be expressed as either a number
+ * or a group name.
+ */
+static int
+set_group(const char * opt_group)
+{
+	gid_t gid;
+	struct group * group_info;
+	uintmax_t parsed;
+
+	/* Find numerical GID. */
+	if ((opt_group[0] >= '0') && (opt_group[0] <= '9')) {
+		parsed = strtoumax(opt_group, NULL, 0);
+		/*
+		 * POSIX says that "uid_t, gid_t... shall be integer types",
+		 * so they might be signed or unsigned.  Since we can't assume
+		 * it one way or the other, we simply pass the value to gid.
+		 */
+		gid = (gid_t)parsed;
+	} else {
+		if ((group_info = getgrnam(opt_group)) == NULL) {
+			warnp("Failed to find group");
+			goto err0;
+		}
+		gid = group_info->gr_gid;
+	}
+
+	/* Set GID. */
+	if (setgid(gid) != 0) {
+		warnp("Failed to set GID");
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * set_user(user_string):
+ * Sets the UID to ${group_string}, which may be expressed as either a number
+ * or a user name.
+ */
+static int
+set_user(const char * opt_user)
+{
+	uid_t uid;
+	struct passwd * user_info;
+	uintmax_t parsed;
+
+	/* Calculate numerical UID. */
+	if ((opt_user[0] >= '0') && (opt_user[0] <= '9')) {
```

Comment:
Does POSIX say that you can't have a user name which starts with a digit?  Don't be systemd.

Also, why not use parsenum here?  It will correctly handle `[ug]id_t` regardless of whether they are signed or unsigned.  If parsenum fails, it means the string is not completely numeric, so you can throw it at getpwnam at that point.

Reply:
Do we actually want to support people specifying a UID directly?  At first I thought we should, but now I'm thinking that it opens a bag of worms.  Let's merely allow people to enter the username.

Apparently on Ubuntu 14.04 at least, an all-digit username of `232` is fine.  And for bonus lols, he even made a username of U+1F600 (😀)!
https://unix.stackexchange.com/questions/287077/why-cant-linux-usernames-begin-with-numbers/287086#287086

Reply:
(the smiley-face username is a POSIX violation, but still awesome.)

The actual POSIX rules allow for digit-only usernames.    The username is composed of characters from the "portable filenames set" (alphanum, plus `.` `-` `_`)  (huh, no other punctuation is deemed to be portable for filenames?).  The only other note given is that the username should not begin with the hyphen character.